metashare/api/views.py

```diff
@@ -107,3 +109,50 @@ class ScratchOrgViewSet(viewsets.ModelViewSet):
 
     def perform_destroy(self, instance):
         instance.queue_delete()
+
+    def list(self, request, *args, **kwargs):
+        # XXX: This method is copied verbatim from
+        # rest_framework.mixins.RetrieveModelMixin, because I needed to
+        # insert the get_unsaved_changes line in the middle.
+        queryset = self.filter_queryset(self.get_queryset())
+
+        # XXX: I am apprehensive about the possibility of flooding the
+        # worker queues easily this way:
+        for instance in queryset.filter(org_type=SCRATCH_ORG_TYPES.Dev):
+            instance.get_unsaved_changes()
```

Comment:
@wlonk So. You're correct here. A few page-refreshes can very quickly lead to a backlog of essentially-duplicate jobs in the queue. Some thoughts (but I'm not sure how technically feasible they are):

- If there's already a job running for that scratch_org, I think we can skip this. Practically I think that includes a `delete`, `create`, `commit`, or another `update` job. That is, if an org is actively being provisioned, deleted, capturing-changes, or already updating-unsaved-changes, I don't think we want to queue another update for that same org.
- If there's not a convenient automated way to do that (e.g. "hey there's another job running for this org, let's bail"), I think we could effectively do it by only triggering this `instance.get_unsaved_changes()` in the following cases:
  - `instance.org_type == SCRATCH_ORG_TYPES.Dev`
  - `instance.url != None`
  - `instance.delete_queued_at == None`
  - `instance.currently_capturing_changes == False`
  - `instance.currently_refreshing_changes == False`

Reply:
Research indicates that the thing to do is a lock (perhaps in Redis, like we did for MetaDeploy), where we use some unique natural key (the args passed to the job, for instance) and set the lock when we enqueue it, and unset it when the job completes or after a timeout. (The timeout we can do in-redis, I think, with setting an expiration on the key.)

Reply:
But I think we can use your option 2 more simply so I'd rather do that.